openedx_tagging/core/tagging/rest_api/v1/serializers.py

```diff
@@ -192,3 +195,70 @@ def get_children_count(self, obj):
         Returns the number of child tags of the given tag.
         """
         return len(obj.sub_tags)
+
+
+class TaxonomyTagCreateBodySerializer(serializers.Serializer):  # pylint: disable=abstract-method
+    """
+    Serializer of the body for the Taxonomy Tags CREATE view
+    """
+
+    tag = serializers.CharField(required=True)
+    parent_tag_value = serializers.CharField(
+        source='parent.value', required=False
+    )
+    external_id = serializers.CharField(required=False)
+
+    def validate_parent_tag_value(self, value):
+        """
+        Check that the provided parent Tag exists based on the value
+        """
+        valid = Tag.objects.filter(value__iexact=value).exists()
+        if not valid:
+            raise serializers.ValidationError("Invalid `parent_tag_value` provided")
+
+        return value
+
+
+class TaxonomyTagUpdateBodySerializer(serializers.Serializer):  # pylint: disable=abstract-method
+    """
+    Serializer of the body for the Taxonomy Tags UPDATE view
+    """
+
+    tag = serializers.CharField(source="value", required=True)
+    updated_tag_value = serializers.CharField(required=True)
+
+    def validate_tag(self, value):
+        """
+        Check that the provided Tag exists based on the value
+        """
+
+        valid = Tag.objects.filter(value__iexact=value).exists()
+        if not valid:
+            raise serializers.ValidationError("Invalid `tag` provided")
+
+        return value
+
+
+class TaxonomyTagDeleteBodySerializer(serializers.Serializer):  # pylint: disable=abstract-method
+    """
+    Serializer of the body for the Taxonomy Tags DELETE view
+    """
+
+    tags = serializers.ListField(
+        child=serializers.CharField(), required=True
+    )
+    with_subtags = serializers.BooleanField(required=False)
+
+    def validate_tags(self, value):
+        """
+        Check that the provided Tags exists based on the values
+        """
+
+        # Build query to check if case-insensitive value for Tags list provided
+        q_list = map(lambda n: Q(value__iexact=n), value)
+        q_list = reduce(lambda a, b: a | b, q_list)
+        valid = Tag.objects.filter(q_list).count() == len(value)
```

Comment:
The implementation in the `Taxonomy.delete_tags` is actually an incorrect one, since the simple `value__in` filter doesnt account for case-insensitve values, the current tests didn't catch it. Thanks for pointing it out, I'll update it.

> Also, do we need to this extra validation in these serializers, since the implementation also enforces the same rules?

My initial thought process was to add the validation in the api incase the functions were called directly rather than only in the rest_api/views (through serializers), however then I figured we could also make use of the validation that DRF provides in addition to that.

The rules they enforce are _slightly_ different, you'll notice the validation that happens in the serializer checks if the Tag values provided are valid (i.e exist in the DB regardless to which taxonomy they belong to) however the validation that happens in the api method checks if the **Taxonomy** has the provided (Tag) values, slight difference where the former would return a 400 and the later would return a 404. I guess the end result is technically the same (the request fails), just thought this give more granularity in checks and errors.

Reply:
> The implementation in the Taxonomy.delete_tags is actually an incorrect one, since the simple value__in filter doesnt account for case-insensitve values, the current tests didn't catch it.

Actually, doing some tests, surprising the simple implementation works 😮, It might be because of the special custom `case_insensitive_char_field` we use in the model? Since that is generally not the default behavior for django's charfields. In any case, that was a pleasant surprise!

Reply:
Ah lovely, I'm glad to hear that case-insensitive fields aren't so hard to interact with!

I'm still not convinced that the extra Serializer validation is needed, since the model does what it does and more. But I'll leave that to you & upstream's discretion.